lib/rules/semi-style.js

```diff
@@ -0,0 +1,131 @@
+        const forLoopHeadLoc = typeof options === "string" ? options : (options.forLoopHead || "last");
+
+        /**
+         * Check whether comments exist between the given 2 tokens.
+         * @param {Token} left The left token to check.
+         * @param {Token} right The right token to check.
+         * @returns {boolean} `true` if comments exist between the given 2 tokens.
+         */
+        function commentsExistBetween(left, right) {
+            return sourceCode.getFirstTokenBetween(
+                left,
+                right,
+                {
+                    includeComments: true,
+                    filter: astUtils.isCommentToken
+                }
+            ) !== null;
+        }
+
+        /**
+         * Check the given semicolon token.
+         * @param {Token} semiToken The semicolon token to check.
+         * @param {string} expected The expected location to check.
+         * @returns {void}
+         */
+        function check(semiToken, expected) {
+            const prevToken = sourceCode.getTokenBefore(semiToken);
+            const nextToken = sourceCode.getTokenAfter(semiToken);
+            const prevIsSameLine = !prevToken || astUtils.isTokenOnSameLine(prevToken, semiToken);
+            const nextIsSameLine = !nextToken || astUtils.isTokenOnSameLine(semiToken, nextToken);
+
+            if ((expected === "last" && !prevIsSameLine) || (expected === "first" && !nextIsSameLine)) {
+                context.report({
+                    loc: semiToken.loc,
+                    message: "Expected this semicolon to be at {{pos}}.",
+                    data: {
+                        pos: (expected === "last")
+                            ? "the end of the previous line"
+                            : "the beginning of the next line"
+                    },
+                    fix(fixer) {
+                        if (commentsExistBetween(prevToken, nextToken)) {
+                            return null;
+                        }
+
+                        const start = prevToken ? prevToken.range[1] : semiToken.range[0];
+                        const end = nextToken ? nextToken.range[0] : semiToken.range[1];
+                        const text = (expected === "last") ? ";\n" : "\n;";
+
+                        return fixer.replaceTextRange([start, end], text);
+                    }
+                });
+            }
+        }
+
+        return {
+            [SELECTOR](node) {
+                const lastToken = sourceCode.getLastToken(node);
+
+                if (astUtils.isSemicolonToken(lastToken)) {
+                    check(lastToken, statementLoc);
+                }
+            },
+
+            ForStatement(node) {
+                const firstSemi = node.init && sourceCode.getTokenAfter(node.init);
+                const secondSemi = node.test && sourceCode.getTokenAfter(node.test);
```

Comment:
I think this will be incorrect if `node.init` or `node.test` is parenthesized. The following code should be valid, but the rule reports an error:

```js
/* eslint semi-style: [error, "first"] */
for (
  (i = 0)
  ; i < 10
  ; i++
) {}
```